## Part 1: Preprocessing

In [187]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
import sklearn as skl
import tensorflow as tf
from tensorflow.keras.models import Model
from tensorflow.keras import layers
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df

Age Attrition     BusinessTravel              Department  \
0      41       Yes      Travel_Rarely                   Sales   
1      49        No  Travel_Frequently  Research & Development   
2      37       Yes      Travel_Rarely  Research & Development   
3      33        No  Travel_Frequently  Research & Development   
4      27        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
1465   36        No  Travel_Frequently  Research & Development   
1466   39        No      Travel_Rarely  Research & Development   
1467   27        No      Travel_Rarely  Research & Development   
1468   49        No  Travel_Frequently                   Sales   
1469   34        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                    1          2  Life Sciences                        2   
1                    8          1  Life Sciences                        3   
2                    2          2          Other                        4   
3                    3          4  Life Sciences                        4   
4                    2          1        Medical                        1   
...                ...        ...            ...                      ...   
1465                23          2        Medical                        3   
1466                 6          1        Medical                        4   
1467                 4          3  Life Sciences                        2   
1468                 2          3        Medical                        4   
1469                 8          3        Medical                        2   

      HourlyRate  JobInvolvement  ...  PerformanceRating  \
0             94               3  ...                  3   
1             61               2  ...                  4   
2             92               2  ...                  3   
3             56               3  ...                  3   
4             40               3  ...                  3   
...          ...             ...  ...                ...   
1465          41               4  ...                  3   
1466          42               2  ...                  3   
1467          87               4  ...                  4   
1468          63               2  ...                  3   
1469          82               4  ...                  3   

     RelationshipSatisfaction  StockOptionLevel TotalWorkingYears  \
0                           1                 0                 8   
1                           4                 1                10   
2                           2                 0                 7   
3                           3                 0                 8   
4                           4                 1                 6   
...                       ...               ...               ...   
1465                        3                 1                17   
1466                        1                 1                 9   
1467                        2                 1                 6   
1468                        4                 0                17   
1469                        1                 0                 6   

      TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                         0               1               6   
1                         3               3              10   
2                         3               3               0   
3                         3               3               8   
4                         3               3               2   
...                     ...             ...             ...   
1465                      3               3               5   
1466                      5               3               7   
1467                      0               3               6   
1468                      3               2               9   
1469                      3               4  

In [188]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [189]:
attrition_df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [190]:
# Encode OverTime
le_O = LabelEncoder()
attrition_df['OverTime'] = le_O.fit_transform(attrition_df['OverTime'])


In [191]:
# Encode Married

le_M = LabelEncoder()
attrition_df['MaritalStatus'] = le_M.fit_transform(attrition_df['MaritalStatus'])

In [192]:
# Encode Attrition
le_A = LabelEncoder()
attrition_df['Attrition'] = le_A.fit_transform(attrition_df['Attrition'])

In [193]:
# Encode Department(LabelEncoder)
le_D = LabelEncoder()
attrition_df['Department'] = le_D.fit_transform(attrition_df['Department'])

In [194]:
attrition_df

Age  Attrition     BusinessTravel  Department  DistanceFromHome  \
0      41          1      Travel_Rarely           2                 1   
1      49          0  Travel_Frequently           1                 8   
2      37          1      Travel_Rarely           1                 2   
3      33          0  Travel_Frequently           1                 3   
4      27          0      Travel_Rarely           1                 2   
...   ...        ...                ...         ...               ...   
1465   36          0  Travel_Frequently           1                23   
1466   39          0      Travel_Rarely           1                 6   
1467   27          0      Travel_Rarely           1                 4   
1468   49          0  Travel_Frequently           2                 2   
1469   34          0      Travel_Rarely           1                 8   

      Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0             2  Life Sciences                        2          94   
1             1  Life Sciences                        3          61   
2             2          Other                        4          92   
3             4  Life Sciences                        4          56   
4             1        Medical                        1          40   
...         ...            ...                      ...         ...   
1465          2        Medical                        3          41   
1466          1        Medical                        4          42   
1467          3  Life Sciences                        2          87   
1468          3        Medical                        4          63   
1469          3        Medical                        2          82   

      JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0                  3  ...                  3                        1   
1                  2  ...                  4                        4   
2                  2  ...                  3                        2   
3                  3  ...                  3                        3   
4                  3  ...                  3                        4   
...              ...  ...                ...                      ...   
1465               4  ...                  3                        3   
1466               2  ...                  3                        1   
1467               4  ...                  4                        2   
1468               2  ...                  3                        4   
1469               4  ...                  3                        1   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

      WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                   1               6                   4   
1                   3              10                   7   
2                   3               0                   0   
3                   3               8                   7   
4                   3               2                   2   
...               ...             ...                 ...   
1465                3               5                   2   
1466                3               7                   7   
1467                3            

In [195]:
# Encode Business Travel 
le_B = LabelEncoder()
attrition_df['BusinessTravel'] = le_B.fit_transform(attrition_df['BusinessTravel'])

In [196]:
# Encode EducationField 
le_E = LabelEncoder()
attrition_df['EducationField'] = le_E.fit_transform(attrition_df['EducationField'])

In [197]:
#Encode JobRole
le_J = LabelEncoder()
attrition_df['JobRole'] = le_J.fit_transform(attrition_df['JobRole'])

In [198]:
attrition_df

Age  Attrition  BusinessTravel  Department  DistanceFromHome  Education  \
0      41          1               2           2                 1          2   
1      49          0               1           1                 8          1   
2      37          1               2           1                 2          2   
3      33          0               1           1                 3          4   
4      27          0               2           1                 2          1   
...   ...        ...             ...         ...               ...        ...   
1465   36          0               1           1                23          2   
1466   39          0               2           1                 6          1   
1467   27          0               2           1                 4          3   
1468   49          0               1           2                 2          3   
1469   34          0               2           1                 8          3   

      EducationField  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
0                  1                        2          94               3   
1                  1                        3          61               2   
2                  4                        4          92               2   
3                  1                        4          56               3   
4                  3                        1          40               3   
...              ...                      ...         ...             ...   
1465               3                        3          41               4   
1466               3                        4          42               2   
1467               1                        2          87               4   
1468               3                        4          63               2   
1469               3                        2          82               4   

      ...  PerformanceRating  RelationshipSatisfaction  StockOptionLevel  \
0     ...                  3                         1                 0   
1     ...                  4                         4                 1   
2     ...                  3                         2                 0   
3     ...                  3                         3                 0   
4     ...                  3                         4                 1   
...   ...                ...                       ...               ...   
1465  ...                  3                         3                 1   
1466  ...                  3                         1                 1   
1467  ...                  4                         2                 1   
1468  ...                  3                         4                 0   
1469  ...                  3                         1                 0   

      TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
0                     8                      0                1   
1                    10                      3                3   
2                     7                      3                3   
3                     8                      3                3   
4                     6                      3                3   
...                 ...                    ...              ...   
1465                 17                      3                3   
1466                  9                      5                3   
1467                  6                      0                3   
1468                 17                      3                2   
1469                  6                      3                4   

      YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0                  6                   4                        0   
1                 10                   7                        1   
2                  0                   0                        0   
3                  8                   7                        3   
4                  2                   2     

In [199]:
attrition_corr = attrition_df.corr()['Attrition'].sort_values(ascending=False)    
attrition_corr

Attrition                   1.000000
OverTime                    0.246118
MaritalStatus               0.162070
DistanceFromHome            0.077924
JobRole                     0.067151
Department                  0.063991
NumCompaniesWorked          0.043494
EducationField              0.026846
PerformanceRating           0.002889
BusinessTravel              0.000074
HourlyRate                 -0.006846
PercentSalaryHike          -0.013478
Education                  -0.031373
YearsSinceLastPromotion    -0.033019
RelationshipSatisfaction   -0.045872
TrainingTimesLastYear      -0.059478
WorkLifeBalance            -0.063939
EnvironmentSatisfaction    -0.103369
JobSatisfaction            -0.103481
JobInvolvement             -0.130016
YearsAtCompany             -0.134392
StockOptionLevel           -0.137145
YearsWithCurrManager       -0.156199
Age                        -0.159205
YearsInCurrentRole         -0.160545
JobLevel                   -0.169105
TotalWorkingYears          -0.171063
N

In [200]:
# Create y_df with the Attrition and Department columns

y_df = attrition_df[['Attrition', 'Department']]

In [201]:
y_df['Attrition'].value_counts()

Attrition
0    1233
1     237
Name: count, dtype: int64

In [202]:
y_df['Department'].value_counts()

Department
1    961
2    446
0     63
Name: count, dtype: int64

In [203]:
# The Attrition and Department parameters are not balanced.

In [204]:
# Create a list of at least 10 column names to use as X data

X_selected = ['OverTime', 'MaritalStatus', 'JobInvolvement', 'YearsAtCompany', 'StockOptionLevel', 'YearsWithCurrManager', 'Age', 'YearsInCurrentRole', 'JobLevel', 'TotalWorkingYears']

# Create X_df using your selected columns
X_df = attrition_df[X_selected]

# Show the data types for X_df
X_df.dtypes


OverTime                int64
MaritalStatus           int64
JobInvolvement          int64
YearsAtCompany          int64
StockOptionLevel        int64
YearsWithCurrManager    int64
Age                     int64
YearsInCurrentRole      int64
JobLevel                int64
TotalWorkingYears       int64
dtype: object

In [205]:
X_df['OverTime'].value_counts()

OverTime
0    1054
1     416
Name: count, dtype: int64

In [206]:
#Collect non-numerics

non_num_df = X_df.select_dtypes(include=object)

for column in non_num_df.columns:
    print(non_num_df[column].value_counts())

In [207]:
non_num_df.columns

Index([], dtype='object')

In [208]:
X_df

OverTime  MaritalStatus  JobInvolvement  YearsAtCompany  \
0            1              2               3               6   
1            0              1               2              10   
2            1              2               2               0   
3            1              1               3               8   
4            0              1               3               2   
...        ...            ...             ...             ...   
1465         0              1               4               5   
1466         0              1               2               7   
1467         1              1               4               6   
1468         0              1               2               9   
1469         0              1               4               4   

      StockOptionLevel  YearsWithCurrManager  Age  YearsInCurrentRole  \
0                    0                     5   41                   4   
1                    1                     7   49                   7   
2                    0                     0   37                   0   
3                    0                     0   33                   7   
4                    1                     2   27                   2   
...                ...                   ...  ...                 ...   
1465                 1                     3   36                   2   
1466                 1                     7   39                   7   
1467                 1                     3   27                   2   
1468                 0                     8   49                   6   
1469                 0                     2   34                   3   

      JobLevel  TotalWorkingYears  
0            2                  8  
1            2                 10  
2            1                  7  
3            1                  8  
4            1                  6  
...        ...                ...  
1465         2                 17  
1466         3                  9  
1467         2                  6  
1468         2                 17  
1469         2                  6  

[1470 rows x 10 columns]

In [209]:
y_df['Department'].value_counts()

Department
1    961
2    446
0     63
Name: count, dtype: int64

In [210]:
y_df['Attrition'].value_counts()

Attrition
0    1233
1     237
Name: count, dtype: int64

In [211]:
department_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore', drop=None)

department_encoder.fit(y_train[['Department']])

y_department_train = department_encoder.transform(y_train[['Department']])
y_department_test = department_encoder.transform(y_test[['Department']])

y_department_train, y_department_test



(array([[0., 1., 0.],
        [0., 0., 1.],
        [0., 1., 0.],
        ...,
        [0., 0., 1.],
        [0., 0., 1.],
        [0., 1., 0.]]),
 array([[0., 1., 0.],
        [0., 0., 1.],
        [0., 0., 1.],
        ...,
        [0., 0., 1.],
        [0., 1., 0.],
        [0., 1., 0.]]))

In [212]:
attrition_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore', drop=None)

attrition_encoder.fit(y_train[['Attrition']])

y_attr_train = attrition_encoder.transform(y_train[['Attrition']])
y_attr_test = attrition_encoder.transform(y_test[['Attrition']])

y_attr_train, y_attr_test

(array([[1., 0.],
        [1., 0.],
        [1., 0.],
        ...,
        [1., 0.],
        [1., 0.],
        [1., 0.]]),
 array([[0., 1.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 

In [213]:
X_train, X_test, y_department_train, y_department_test, y_attr_train, y_attr_test = train_test_split(X_df, y_df['Department'], y_df['Attrition'])

In [214]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
scaler.fit(X_train)
# Scale the training and testing data

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Create, Compile, and Train the Model

In [215]:
X_df.shape

(1470, 10)

In [216]:
# Create the shared layers of the model

# Input layer
input_layer = layers.Input(shape=(X_df.shape[1],), name='input_features')

# Shared hidden layers
shared_layer1 = layers.Dense(64, activation='relu')(input_layer)
shared_layer2 = layers.Dense(32, activation='relu')(shared_layer1)

In [217]:
# Branch for department prediction
y_department_output = layers.Dense(3, activation='softmax', name='y_department_output')(shared_layer2)

# Branch for attrition prediction
y_attr_output = layers.Dense(1, activation='sigmoid', name='y_attr_output')(shared_layer2)

In [218]:
# Create the model
model = Model(inputs=input_layer, outputs=[y_department_output, y_attr_output])

# Compile the model
model.compile(optimizer='adam',
            loss={'department_output': 'categorical_crossentropy', 'attr_output': 'binary_crossentropy'},
            metrics={'department_output': 'accuracy', 'attr_output': 'accuracy'})

# Display the model summary
model.summary()

Model: "functional_16"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_features      │ (None, 10)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_11 (Dense)    │ (None, 64)        │        704 │ input_features[0… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_12 (Dense)    │ (None, 32)        │      2,080 │ dense_11[0][0]    │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ y_department_output │ (None, 3)         │         99 │ dense_12[0][0]    │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ y_attr_output       │ (None, 1)         │         33 │ dense_12[0][0]    │
│ (Dense)             │                   │            │                   │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 2,916 (11.39 KB)

 Trainable params: 2,916 (11.39 KB)

 Non-trainable params: 0 (0.00 B)

In [219]:
# Fit the model
model.fit(
    X_df,
    {'y_department_output': y_department_output, 'y_attr_output': y_attr_output},
    epochs=10,
    batch_size=32,
    validation_split=0.2
)

ValueError: A KerasTensor is symbolic: it's a placeholder for a shape an a dtype. It doesn't have any actual numerical value. You cannot convert it to a NumPy array.

In [ ]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_test_scaled, {'y_department_output': y_department_test, 'y_attr_output': y_attr_test})
test_results

ValueError: In the dict argument `loss`, key 'department_output' does not correspond to any model output. Received:
loss={'department_output': 'categorical_crossentropy', 'attr_output': 'binary_crossentropy'}

In [ ]:
# Print the quality and color accuracy
print(f"Department Accuracy: {test_results[0]}")
print(f"Attrition Accuracy: {test_results[1]}")


In [ ]:
# # Create a OneHotEncoder for the Department column
# input_nodes = len(X_df.columns)

# # Fit the encoder to the training data
# input_layer = layers.Input(shape=(input_nodes,), name='input_layer')

# # Create two new variables by applying the encoder
# shared1 = layers.Dense(64, activation='relu')(input_layer)
# shared2 = layers.Dense(128, activation='relu')(shared1)
# # to the training and testing data

In [ ]:
# Department_hidden = layers.Dense(32, activation='relu')(shared2)

# Department_output = layers.Dense(3, activation='softmax', name='Department_out')(Department_hidden)

In [ ]:
# # Create attrition branch
# # Create the hidden layer
# Attrition_hidden = layers.Dense(32, activation='relu')(shared2)

# # Create the output layer
# Attrition_output = layers.Dense(2, activation='sigmoid', name='Attrition_out')(Attrition_hidden)

In [ ]:
# # Create the model
# model = Model(inputs=input_layer, outputs=[Department_output, Attrition_output])

# # Compile
# model.compile(optimizer='adam',
#                 loss={'Department_out': 'categorical_crossentropy', 'Attrition_out': 'binary_crossentropy'},
#                 metrics={'Department_out': 'accuracy', 'Attrition_out': 'accuracy'})

# model.summary()

Model: "functional_3"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 10)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_4 (Dense)     │ (None, 64)        │        704 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_5 (Dense)     │ (None, 128)       │      8,320 │ dense_4[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_6 (Dense)     │ (None, 32)        │      4,128 │ dense_5[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_7 (Dense)     │ (None, 32)        │      4,128 │ dense_5[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Department_out      │ (None, 3)         │         99 │ dense_6[0][0]     │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Attrition_out       │ (None, 2)         │         66 │ dense_7[0][0]     │
│ (Dense)             │                   │            │                   │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 17,445 (68.14 KB)

 Trainable params: 17,445 (68.14 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
# # Train the model
# model.fit(X_train_scaled,
#             {'Department_out': y_department_train, 'Attrition_out': y_attr_train},
#             epochs=100,
#             batch_size=32,
#             validation_split=0.2)

Epoch 1/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - Attrition_out_accuracy: 0.8135 - Department_out_accuracy: 0.5440 - loss: 1.5560 - val_Attrition_out_accuracy: 0.7828 - val_Department_out_accuracy: 0.6833 - val_loss: 1.2952
Epoch 2/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_out_accuracy: 0.8512 - Department_out_accuracy: 0.6315 - loss: 1.2315 - val_Attrition_out_accuracy: 0.7828 - val_Department_out_accuracy: 0.6923 - val_loss: 1.2413
Epoch 3/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_out_accuracy: 0.8552 - Department_out_accuracy: 0.6359 - loss: 1.1327 - val_Attrition_out_accuracy: 0.7919 - val_Department_out_accuracy: 0.6878 - val_loss: 1.2363
Epoch 4/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_out_accuracy: 0.8972 - Department_out_accuracy: 0.6461 - loss: 1.0538 - val_Attrition_out_accuracy: 0.7873 - val_Department_out_accuracy: 0.6923 - val_loss: 1.2280
Epoch 5/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_out_accuracy: 0.8715 - Depart

In [ ]:
model.metrics_names

['loss', 'compile_metrics']

In [ ]:
# # Evaluate the model
# test_results = model.evaluate(X_test_scaled, {'Department_out': y_department_test, 'Attrition_out': y_attr_test})
# print(test_results)

12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 722us/step - Attrition_out_accuracy: 0.8169 - Department_out_accuracy: 0.5767 - loss: 2.9810
[2.964831590652466, 0.804347813129425, 0.5869565010070801]


In [ ]:
# # Print the accuracy for both department and attrition
# print(f"Attrition predictions accuracy: {test_results[1]:.3f}")
# print(f"Department predictions accuracy: {test_results[2]:.3f}")

Attrition predictions accuracy: 0.804
Department predictions accuracy: 0.587


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?
     
2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

# Summary
YOUR ANSWERS HERE

1.  Not very accurate.  For instance, Department predictions accuracy is only around 58% versus Attrition accuracy  
    83%. 
2.  The activation function I chose was relu for hidden layers and sigmoid for the output. 
3.  I could have tried to picked a different X parameters to try to see if that improves.